docs/DafnyRef/UserGuide.md

```diff
@@ -93,10 +93,43 @@ However, which files are verified, built, run, or processed by other
 dafny commands depends on the individual command. 
 These commands are described in [Section 13.5.1](#sec-dafny-commands).
 
-
 [^fn-duplicate-files]: Files may be included more than once or both included and listed on the command line. Duplicate inclusions are detected and each file processed only once.
 For the purpose of detecting duplicates, file names are considered equal if they have the same absolute path, compared as case-sensitive strings (regardless of whether the underlying file-system is case sensitive).  Using symbolic links may make the same file have a different absolute path; this will generally cause duplicate declaration errors.
 
+### 13.3.1. Dafny Build Artifacts: the Library Backend and .doo Files {#sec-doo-files}
+
+As of Dafny 4.1, `dafny` now supports outputting a single file containing
+a fully-verified program along with metadata about how it was verified.
+Such files use the extension `.doo`, for Dafny Output Object,
+and can be used as input anywhere a `.dfy` file can be.
+
+`.doo` files are produced by an additional backend called the "Dafny Library" backend,
+identified with the name `lib` on the command line. For example, to build multiple
+Dafny files into a single build artifact for shared reuse, the command would look something like:
+
+```bash
+dafny build -t:lib A.dfy B.dfy C.dfy --out MyLib.doo
```

Comment:
I think Dafny projects should specify whether they're an executable or a library, and depending on this, the _default_ for `dafny build` should produce an executable or a `*.doo` file.

Reply:
> I think Dafny projects should specify whether they're an executable or a library, and depending on this, the _default_ for `dafny build` should produce an executable or a `*.doo` file.
> 
> Also, I think the "lib" target should not exist for `dafny run` and `dafny translate`. Currently, it's not mentioned in the `--help` documentation for those (and for build).
> 
> I think the ideal end-game is that Dafny's standard library is large enough that `dafny run` and `dafny build` can generally be used without supplying additional non-Dafny files that implement externs, and that `dafny translate` is used for integration scenarios. If that's the case, the `--target` option for `dafny run` and `build` would generally not be used, and `*.doo` files would be generated when doing `dafny build` on a Dafny project that's marked as a library.
> 
> For now, could you show an error when using the `lib` target with `run`, and have `lib` be the default for projects without a main method?

You make some good points, but I think the real issue is us using "lib" as an identifier for the ".doo target", because it implies the result is a library and not an executable, when that aspect should be orthogonal to which backend you're targeting. We could just call it "the doo backend": I didn't love that, but also couldn't think of a better label.

I'd much rather have the .doo backend support all operations, so that it's easier to get better coverage through generic utilities like `%testDafnyForEachCompiler`. Hopefully changing the label would make that seem more reasonable.

> Currently I get:
> 
> ```
> rwillems@bcd0745419f2 dafny2 % ./Scripts/dafny run -t:lib /Users/rwillems/SourceCode/dafny2/Test/c++/hello.dfy                
> 
> Dafny program verifier finished with 0 verified, 0 errors
> Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
>    at DafnyCore.DooFile.Validate(String filePath, DafnyOptions options, Command currentCommand) in /Users/rwillems/SourceCode/dafny2/Source/DafnyCore/DooFile.cs:line 122
>    at Microsoft.Dafny.DafnyFile..ctor(DafnyOptions options, String filePath, Boolean useStdin) in /Users/rwillems/SourceCode/dafny2/Source/DafnyCore/DafnyMain.cs:line 99
>    at Microsoft.Dafny.DafnyDriver.ProcessCommandLineArguments(String[] args, DafnyOptions& options, List`1& dafnyFiles, List`1& otherFiles) in /Users/rwillems/SourceCode/dafny2/Source/DafnyDriver/DafnyDriver.cs:line 263
>    at Microsoft.Dafny.DafnyDriver.ThreadMain(String[] args) in /Users/rwillems/SourceCode/dafny2/Source/DafnyDriver/DafnyDriver.cs:line 112
>    at Microsoft.Dafny.DafnyDriver.<>c__DisplayClass10_0.<Main>b__0() in /Users/rwillems/SourceCode/dafny2/Source/DafnyDriver/DafnyDriver.cs:line 102
>    at System.Threading.Thread.StartCallback()
>  Process exited with exit code 134
> ```
> 
> and
> 
> ```
> rwillems@bcd0745419f2 dafny2 % ./Scripts/dafny translate lib /Users/rwillems/SourceCode/dafny2/Test/c++/hello.dfy 
> 
> Dafny program verifier finished with 0 verified, 0 errors
> ```

The first case is a legit bug which I've got a fix for, thanks for finding that: https://github.com/dafny-lang/dafny/pull/3967

The second case is working as advertised AFAICT: it should have created a `hello-lib` directory with the entries for a doo file (which I don't love as discussed above, but I didn't think it was actively harmful, and I found it handy for debugging).

Reply:
OK, I like your view of things. We should change the `TargetId` `lib` into `doo` or `dfy` though.

Reply:
Based on offline discussions with @robin-aws , created this issue: https://github.com/dafny-lang/dafny/issues/3976